### PR TITLE
[JUJU-3969] Allow upgrades without specifying version

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following allows migrating to a new version of juju without explicitly stating the newer major version (3.x -> 4.x). Juju shouldn't have to know about new major versions, so to fix this requires removing the code that was recently added to pass the target controller version to the source pre-checks. In addition to this, I've also refactored the migration precheck context types to not allow target pre-check to explicitly know about model checking. This should prevent any target and source issues.

This is required as 3.1 was earmarked as the last LTS before releasing 4.0. The target controller will then validate where the model is coming from, and ensure it's within its own range before allowing it to migrate.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

We require this [PR](https://github.com/juju/juju/pull/15715) one to land before we can fully test the workflow.

To ensure we didn't break any backward compatibility:

```sh
$ sudo snap install juju --channel 3.0/stable
$ juju bootstrap lxd test30
$ juju add-model default
$ juju deploy ubuntu
$ sudo snap remove juju --purge
$ make go-install
$ juju bootstrap lxd test31
$ juju switch controller
$ juju wait-for application controller
$ juju switch test30
$ juju wait-for application ubuntu
$ sleep 10
$ juju migrate default test31
$ juju switch test31
$ sleep 10
$ juju switch default
$ juju wait-for application ubuntu 
```

### CODE REVIEW

File: upgrades/upgradevalidation/version_test.go [hunk 1]
```diff
@@ -24,7 +24,6 @@ type versionCheckTC struct {
 	allowed bool
 	minVers string
 	err     string
-	patch   bool
```
Comment (SimonRichardson) [inline upgrades/upgradevalidation/version_test.go:27]: Patch was removed, because it was set to true for every test.

**Review by hmlanigan — DISMISSED** [review]

**Comment by hmlanigan** [comment]
@SimonRichardson what has changed since we last talked? If this will be done in pieces, what are they?

File: upgrades/upgradevalidation/upgrade.go [hunk 2]
```diff
@@ -18,7 +18,7 @@ func ValidatorsForControllerUpgrade(
 			getCheckTargetVersionForControllerModel(targetVersion),
 			checkMongoStatusForControllerUpgrade,
 		}
-		if targetVersion.Major == 3 {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply upgrades/upgradevalidation/upgrade.go:21]: I need to revert this line.

File: upgrades/upgradevalidation/upgrade.go [hunk 3]
```diff
@@ -56,7 +56,7 @@ func ValidatorsForModelUpgrade(
 	validators := []Validator{
 		getCheckUpgradeSeriesLockForModel(force),
 	}
-	if targetVersion.Major == 3 {
```
Comment (SimonRichardson) [inline upgrades/upgradevalidation/upgrade.go:59]: And this line needs to be reverted.

File: upgrades/upgradevalidation/migrate.go [hunk 4]
```diff
@@ -9,23 +9,23 @@ import (
...
 
-// ValidatorsForModelMigrationSource returns a list of validators to run source controller for model migration,
+// ValidatorsForModelMigrationSource returns a list of validators to run source
+// controller for model migration.
+// Note: the target version can never be lower than the current version.
 func ValidatorsForModelMigrationSource(
 	targetVersion version.Number, cloudspec environscloudspec.CloudSpec,
 ) []Validator {
 	validators := []Validator{
 		getCheckTargetVersionForModel(targetVersion, MigrateToAllowed),
```
Comment (SimonRichardson) [inline upgrades/upgradevalidation/migrate.go:19]: I'm still not sure what this gives you, only to block you from moving to new versions.

File: upgrades/upgradevalidation/migrate.go [hunk 5]
```diff
@@ -4,28 +4,20 @@
...
-// ValidatorsForModelMigrationSource returns a list of validators to run source controller for model migration,
+// ValidatorsForModelMigrationSource returns a list of validators to run source
+// controller for model migration.
+// Note: the target version can never be lower than the current version.
 func ValidatorsForModelMigrationSource(
-	targetVersion version.Number, cloudspec environscloudspec.CloudSpec,
+	cloudspec environscloudspec.CloudSpec,
 ) []Validator {
-	validators := []Validator{
-		getCheckTargetVersionForModel(targetVersion, MigrateToAllowed),
```
Comment (SimonRichardson) [inline upgrades/upgradevalidation/migrate.go:17]: This line is the real meat of it. We no longer check the source for the target version. This is totally up to the target to check for.

File: upgrades/upgradevalidation/migrate.go [hunk 6]
```diff
@@ -4,28 +4,20 @@
...
-		validators = append(validators,
-			checkNoWinMachinesForModel,
-			checkForDeprecatedUbuntuSeriesForModel,
-			getCheckForLXDVersion(cloudspec),
-		)
-		if targetVersion.Minor >= 1 {
-			validators = append(validators, checkForCharmStoreCharms)
-		}
-	}
-	return validators
```
Comment (SimonRichardson) [inline upgrades/upgradevalidation/migrate.go:30]: We're already at 3.1 so these branches can just be removed.

File: migration/types.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package migration
```
Comment (wallyworld) [inline migration/types.go:4]: Other packages call this file `interface.go`
Maybe consistency is best?

File: upgrades/upgradevalidation/migrate.go [hunk 8]
```diff
@@ -4,28 +4,20 @@
 package upgradevalidation
 
 import (
-	"github.com/juju/version/v2"
-
 	environscloudspec "github.com/juju/juju/environs/cloudspec"
 )
 
-// ValidatorsForModelMigrationSource returns a list of validators to run source controller for model migration,
+// ValidatorsForModelMigrationSource returns a list of validators to run source
```
Comment (wallyworld) [inline upgrades/upgradevalidation/migrate.go:10]: ... to run on the ...

File: upgrades/upgradevalidation/migrate.go (hunk 5 above)
Comment (wallyworld) [reply upgrades/upgradevalidation/migrate.go:17]: Can `MigrateToAllowed` be deleted?

**Review by wallyworld — COMMENTED** [review]
Is a separate PR doing the version validation on the target controller? We would need to do that up front before any model import is done. We'd lose the advantage of not having to bother to quiesce the source model, unless we added a PreCheck() method on the target controller or something like that.

File: migration/types.go (hunk 7 above)
Comment (SimonRichardson) [reply migration/types.go:4]: Sure.

**Comment by SimonRichardson** [comment]
@wallyworld :point_down: 

> Is a separate PR doing the version validation on the target controller

There is always a check on the target controller. https://github.com/juju/juju/blob/3.2/migration/precheck.go#L258

> Can MigrateToAllowed be deleted?

No, because we need to check this in the target precheck for the source-to-target validation.

This just removes the code that the target was already checking.

**Review by wallyworld — APPROVED** [review]

File: upgrades/upgradevalidation/upgrade.go [hunk 9]
```diff
@@ -9,62 +9,53 @@ import (
...
-		validators = append(validators,
-			checkNoWinMachinesForModel,
-			checkForDeprecatedUbuntuSeriesForModel,
-			getCheckForLXDVersion(cloudspec),
-		)
-		if targetVersion.Minor >= 1 {
-			validators = append(validators, checkForCharmStoreCharms)
-		}
+	if targetVersion.Major == 3 && targetVersion.Minor >= 1 {
+		validators = append(validators, checkForCharmStoreCharms)
```
Comment (hmlanigan) [inline upgrades/upgradevalidation/upgrade.go:58]: I'm not hitting this check. Migration doesn't fail until here:
```machine-0: 20:44:19 ERROR juju.worker.migrationmaster.a98b55 model data transfer failed, failed to migrate binaries: cannot upload charms: cannot upload charm: Post https://10.186.60.233:17070/migrate/charms?arch=&name=ubuntu&revision=24&schema=cs&series=&user=: cannot upload charm: unsupported schema "cs"```

**Review by hmlanigan — CHANGES_REQUESTED** [review]
Problem in QA

File: upgrades/upgradevalidation/upgrade.go (hunk 9 above)
Comment (SimonRichardson) [reply upgrades/upgradevalidation/upgrade.go:58]: Yeah because you're not allowed cs charms in 3.1. So this is correct.

Comment (SimonRichardson) [reply upgrades/upgradevalidation/upgrade.go:58]: So I did some investigation. This is entirely correct for what we have now. If you try and migrate from 2.9 to 3.1 there is no check at all, on the source side.

https://github.com/juju/juju/blob/02d498631e196f2a37f9b7c3b5c31bdcb1dad333/upgrades/upgradevalidation/migrate.go#L13-L28

The target side will only ever validate from 3.1 onwards

https://github.com/juju/juju/blob/d3c8991a8d1be1f5711ae53e096caadeba3244a3/upgrades/upgradevalidation/upgrade.go#L21-L31

So this PR doesn't actually break any existing flow from 2.9 to 3.1. It's currently been this way. There is no way to get a cs charm into 3.1, so we could actually remove the check on the source side as well.

Here are my steps to reproduce exactly the same thing as this comment:

```sh
$ snap install juju --classic --channel=2.9/stable
$ /snap/bin/juju bootstrap lxd src29
$ /snap/bin/juju deploy cs:~jameinel/ubuntu-lite
```

```sh
$ git checkout 3.1
$ juju bootstrap lxd target31 --build-agent
```

```sh
$ /snap/bin/juju migrate src29:default target31
```

@hmlanigan we can fix that issue in a subsequent PR, but not this one, then I'm fine with that.

**Review by hmlanigan — APPROVED** [review]
We can fix the QA fail in a new pr

**Pull request merged**